File: frontend/src/app/modules/tasker-service/service-area/services/service-area.service.ts
import {Injectable} from '@angular/core';
import {ApiService} from "../../../../shared/services/api/api.service";
import {Observable, Subject, takeUntil} from "rxjs";
import {ApiResponseModel} from "../../../../shared/models/api/api-response.model";
import {ApiEndpointConstant} from "../../../../shared/constants/api-endpoint.constant";
import {ServiceAreaModel} from '../models/service-area.model';

@Injectable({
  providedIn: 'root'
})
export class ServiceAreaService extends ApiService {

  public createServiceArea(username: string, serviceAreaModel: ServiceAreaModel): Observable<ApiResponseModel<any>> {
    return this.post<ApiResponseModel<any>>(`${ApiEndpointConstant.TASKER}/${username}${ApiEndpointConstant.SERVICE_AREA}`, serviceAreaModel);
  }

  public updateServiceArea(serviceAreaId: string, serviceAreaModel: ServiceAreaModel): Observable<ApiResponseModel<any>> {
    return this.put<ApiResponseModel<any>>(`${ApiEndpointConstant.SERVICE_AREA}/${serviceAreaId}`, serviceAreaModel);
  }

  public deleteServiceArea(serviceId: string, cancel$: Subject<void>): Observable<ApiResponseModel<any>> {
    return this.delete<ApiResponseModel<any>>(`${ApiEndpointConstant.SERVICE_AREA}/${serviceId}`)
      .pipe(takeUntil(cancel$))
  }

  public getServiceAreaById(serviceAreaId: string): Observable<ApiResponseModel<any>> {
    return this.get<ApiResponseModel<any>>(`${ApiEndpointConstant.SERVICE_AREA}/${serviceAreaId}`);
  }

  public getServiceAreaOptions(username: string): Observable<ApiResponseModel<any>> {
    return this.get<ApiResponseModel<any>>(`${ApiEndpointConstant.TASKER}/${username}${ApiEndpointConstant.SERVICE_AREA}${ApiEndpointConstant.OPTIONS}`);
  }

}

File: backend/src/main/java/com/tasklion/backend/features/taskerService/serviceArea/service/ServiceAreaService.java
package com.tasklion.backend.features.taskerService.serviceArea.service;

import com.tasklion.backend.common.model.PaginationModel;
import com.tasklion.backend.features.taskerService.serviceArea.ServiceArea;
import com.tasklion.backend.features.taskerService.serviceArea.model.ServiceAreaModel;
import com.tasklion.backend.features.taskerService.serviceArea.model.ServiceAreaRequestModel;
import org.springframework.data.domain.Page;

import java.util.List;
import java.util.Map;

public interface ServiceAreaService {

    void deleteServiceArea(Long id);

    Page<ServiceAreaModel> getServiceAreas(String username, PaginationModel paginationModel);

    List<ServiceAreaModel> getServiceAreas(String username);

    Map<String, Map<String, List<String>>> getServiceAreaOptions(String username);

    ServiceAreaModel createServiceArea(String username, ServiceAreaModel serviceAreaModel);

    ServiceAreaModel updateServiceArea(Long id, ServiceAreaModel serviceAreaModel);

    ServiceAreaModel getServiceArea(Long serviceAreaId);

    ServiceArea getServiceArea(ServiceAreaRequestModel serviceAreaRequestModel);

}
